File: frontend/src/api/exchange.ts
import { apiClient } from '@/api/client';

export interface ExchangeContract {
  id: string;
  contract_code: string;
  version: number;
  status: string;
  audience_type: string;
  payload_type: string;
  schema_reference: string | null;
  effective_from: string;
  effective_to: string | null;
  compatibility_notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface Publication {
  id: string;
  building_id: string;
  contract_version_id: string;
  audience_type: string;
  publication_type: string;
  pack_id: string | null;
  content_hash: string;
  published_at: string;
  published_by_org_id: string | null;
  published_by_user_id: string | null;
  delivery_state: string;
  superseded_by_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface ImportReceipt {
  id: string;
  building_id: string | null;
  source_system: string;
  contract_code: string;
  contract_version: number;
  import_reference: string | null;
  imported_at: string;
  status: string;
  content_hash: string;
  rejection_reason: string | null;
  matched_publication_id: string | null;
  notes: string | null;
  created_at: string;
}

export const exchangeApi = {
  async listPublications(buildingId: string): Promise<Publication[]> {
    const res = await apiClient.get<Publication[]>(`/buildings/${buildingId}/passport-publications`);
    return res.data;
  },

  async listImportReceipts(buildingId: string): Promise<ImportReceipt[]> {
    const res = await apiClient.get<ImportReceipt[]>(`/buildings/${buildingId}/import-receipts`);
    return res.data;
  },

  async listContracts(): Promise<ExchangeContract[]> {
    const res = await apiClient.get<ExchangeContract[]>('/exchange/contracts');
    return res.data;
  },
};

File: backend/app/schemas/exchange.py
"""BatiConnect — Exchange contract and publication schemas."""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# --- Exchange Contract Version ---


class ExchangeContractRead(BaseModel):
    id: UUID
    contract_code: str
    version: int
    status: str
    audience_type: str
    payload_type: str
    schema_reference: str | None
    effective_from: date
    effective_to: date | None
    compatibility_notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Passport Publication ---


class PublicationCreate(BaseModel):
    contract_version_id: UUID
    audience_type: str
    publication_type: str
    pack_id: UUID | None = None
    content_hash: str
    delivery_state: str = "published"


class PublicationRead(BaseModel):
    id: UUID
    building_id: UUID
    contract_version_id: UUID
    audience_type: str
    publication_type: str
    pack_id: UUID | None
    content_hash: str
    published_at: datetime
    published_by_org_id: UUID | None
    published_by_user_id: UUID | None
    delivery_state: str
    superseded_by_id: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Import Receipt ---


class ImportReceiptCreate(BaseModel):
    building_id: UUID | None = None
    source_system: str
    contract_code: str
    contract_version: int
    import_reference: str | None = None
    content_hash: str
    status: str = "received"
    rejection_reason: str | None = None
    matched_publication_id: UUID | None = None
    notes: str | None = None


class ImportReceiptRead(BaseModel):
    id: UUID
    building_id: UUID | None
    source_system: str
    contract_code: str
    contract_version: int
    import_reference: str | None
    imported_at: datetime
    status: str
    content_hash: str
    rejection_reason: str | None
    matched_publication_id: UUID | None
    notes: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
